File: rosettacode/Abundant-_deficient_and_perfect_number_classifications/Abundant,_deficient_and_perfect_number_classifications.vb
Module NumberClassification
    Function SumOfDivisors(n As Integer) As Integer
        If n = 1 Then Return 0

        Dim limit = Math.Sqrt(n)
        If limit < 2 Then Return 1

        Return (Aggregate i In Enumerable.Range(2, CInt(limit) - 1)
               Where n Mod i = 0
               Select If(CDbl(i) = limit, i, i + n \ i)
               Into Sum()) + 1
    End Function

    Sub main(args() As String)
        Dim DeficientCount = 0, PerfectCount = 0, AbundantCount = 0

        For n = 1 To 20000
            Dim pn = SumOfDivisors(n)
            If pn < n Then
                DeficientCount += 1
            ElseIf pn = n Then
                PerfectCount += 1
            Else
                AbundantCount += 1
            End If
        Next n

        Console.Write(
            "Deficient integers: {0}" & vbNewLine &
            "Perfect integers: {1}" & vbNewLine &
            "Abundant integers: {2}",
            DeficientCount, PerfectCount, AbundantCount
        )
    End Sub
End Module

File: rosettacode/Abundant-_deficient_and_perfect_number_classifications/Abundant,_deficient_and_perfect_number_classifications.c
#include <stdlib.h>
#include <math.h>
#include <stdio.h>

long long sum_divisors(int n)
{
    int i;
    long long sum = 1;
    double limit;
#if 0
    int ilimit;
#endif

    if (n == 1) return 0;

    limit = sqrt((double) n);
    for (i = 2; i <= limit; i++)
        if (n % i == 0)
            sum += ((double) i == limit)? i : i + n / i;

#if 0
    for (i = 2, ilimit = n / 2; i < ilimit; i++)
    {
        div_t d = div(n, i);
        if (d.rem == 0)
        {
            sum += (i == ilimit)? i : i + d.quot;
            ilimit = d.quot;
        }
    }
#endif

    return sum;
}

const int search_limit = 20000;

int main(void)
{
    int deficient_count = 0,
        perfect_count = 0,
        abundant_count = 0,
        n;

    for (n = 1; n <= search_limit; n++)
    {
        long long pn = sum_divisors(n);
        if (pn > INT_MAX) abundant_count++; /* 'n' is an int, so any 'pn' higher than INT_MAX is clearly abundant */
        else if (pn < n) deficient_count++;
        else if (pn == n) perfect_count++;
        else /* pn > n */ abundant_count++;
    }

    printf("Deficient integers: %d\n"
           "Perfect integers: %d\n"
           "Abundant integers: %d\n",
           deficient_count, perfect_count, abundant_count);

    return EXIT_SUCCESS;
}
